ld-pruning-wf.wdl

```diff
@@ -0,0 +1,437 @@
+	}
+}
+
+# [3] merge_gds -- merge a bunch of GDS files into one
+task merge_gds {
+	input {
+		Array[File] gdss
+
+		# having an input that depends upon a previous task's output reigns in
+		# cromwell's tendencies to run tasks out of order
+		File? enforce_chronological_order
+
+		# optional
+		String? out_prefix
+
+		# runtime attributes
+		Int addldisk = 5
+		Int cpu = 2
+		Int memory = 4
+		Int preempt = 3
+	}
+
+	# Estimate disk size required
+	Int gds_size = ceil(size(gdss, "GB"))
+	Int final_disk_dize = gds_size * 3 + addldisk
+	String filename = select_first([out_prefix, "merged"])
+
+	command <<<
+		set -eux -o pipefail
+
+		# CWL has an ln -s, will probably need to use copy trick again
+		echo "Copying inputs into the workdir"
+		BASH_FILES=(~{sep=" " gdss})
+		for BASH_FILE in ${BASH_FILES[@]};
+		do
+			cp ${BASH_FILE} .
+		done
+
+		echo "Generating config file"
+		python << CODE
+		import os
+
+		def find_chromosome(file):
+			chr_array = []
+			chrom_num = split_on_chromosome(file)
+			if(unicode(str(chrom_num[1])).isnumeric()):
+				# two digit number
+				chr_array.append(chrom_num[0])
+				chr_array.append(chrom_num[1])
+			else:
+				# one digit number or Y/X/M
+				chr_array.append(chrom_num[0])
+			return "".join(chr_array)
+
+		def split_on_chromosome(file):
+			chrom_num = file.split("chr")[1]
+			return chrom_num
+
+		def write_config(chrs, path):
+			f = open("merge_gds.config", "a")
+
+			if "~{out_prefix}" != "":
+				merged_gds_file_name = "~{out_prefix}" + ".gds"
+			else:
+				merged_gds_file_name = "merged.gds"
+
+			f.write('gds_file "' + path[0] + 'chr ' + path[1] + '"\n')
+			f.write('merged_gds_file "' + merged_gds_file_name + '"\n')
+			f.write('chromosomes "' + chrs + '"\n')
+
+			f.close()
+
+		gds_array_fullpath = ['~{sep="','" gdss}']
+		gds_array_basenames = []
+		for fullpath in gds_array_fullpath:
+			gds_array_basenames.append(os.path.basename(fullpath))
+
+		a_file = gds_array_basenames[0]
+		chr = find_chromosome(a_file)
+		path = a_file.split('chr'+chr)
+
+		# make list of all chromosomes found in input files
+		chr_array = []
+		for gds_file in gds_array_basenames:
+			chrom_num = find_chromosome(gds_file)
+			chr_array.append(chrom_num)
+		chrs = ' '.join(chr_array)
+
+		write_config(chrs, path)
+
+		exit()
+		CODE
+
+		Rscript /usr/local/analysis_pipeline/R/merge_gds.R merge_gds.config
+	>>>
+
+	runtime {
+		cpu: cpu
+		disks: "local-disk " + final_disk_dize + " HDD"
+		docker: "uwgac/topmed-master:2.10.0"
+		memory: "${memory} GB"
+		preemptibles: "${preempt}"
+	}
+	
+	output {
+		File merged_gds_output = filename + ".gds"
+		File config_file = "merge_gds.config"
+	}
+}
+
+# [4] check_merged_gds -- check a merged GDS file against its inputs
+
+task check_merged_gds {
+	input {
+		File gds_file
+		File merged_gds_file
+
+		# having an input that depends upon a previous task's output reigns in
+		# cromwell's tendencies to run tasks out of order
+		File? enforce_chronological_order
```

Comment:
How is this getting used?

Reply:
In a standard ld-pruning run, not at all. Each task within ld-prune requires an input from the previous task already, so this isn't necessary.

But the checker workflow becomes a chaotic mess without this. Without this enforcement, it may execute in this order:
default case ld prune --> default case subset --> md5 default case subset --> non-default case ld prune --> default case merge -->non-default case subset -->  md5 non-default case subset --> default case check merge --> non-default case merge -->  md5 default case merge --> md5 non-default case merge --> non-default case check merge

Or dozens of other permutations. I want the checker workflow to go in precisely the order I listed the tasks, because if something breaks, it needs to be clear in what step of the original pipeline it would break. If a user has to go through the same "why did the step AFTER the the failure step even execute, much less succeed?" stuff that I did, that's just not a good experience.

Reply:
My question is: you have a parameter in the task, where in *this* task is it getting used?

Reply:
Nowhere. That's the point. It's a completely bogus file but it manages to trick Cromwell into enforcing an order, because it *thinks* it needs it.

Reply:
We ultimately decided to remove this from the main workflow and only have the bogus input in the checker.